sentry/sentry.go

```diff
@@ -0,0 +1,117 @@
+package sentry
+
+import (
+	"context"
+	"os"
+
+	"github.com/fabric8-services/fabric8-wit/token"
+
+	"github.com/getsentry/raven-go"
+	goajwt "github.com/goadesign/goa/middleware/security/jwt"
+)
+
+// client encapsulates client to Sentry service
+// also has mutex which controls access to the client
+type client struct {
+	c       *raven.Client
+	sendErr chan func()
+}
+
+var (
+	sentryClient *client
+)
+
+// Sentry returns client declared inside package
+func Sentry() *client {
+	return sentryClient
+}
+
+// InitializeSentryClient initializes sentry client. This function returns
+// function that can be used to close the sentry client and error.
+func InitializeSentryClient(options ...func(*client)) (func(), error) {
+	c, err := raven.New(os.Getenv("SENTRY_DSN"))
+	if err != nil {
+		return nil, err
+	}
+	sentryClient = &client{
+		c:       c,
+		sendErr: make(chan func()),
+	}
+	// set all options passed by user
+	for _, opt := range options {
+		opt(sentryClient)
+	}
+
+	// wait on errors to be sent on channel of client object
+	go sentryClient.loop()
+	return func() {
+		close(sentryClient.sendErr)
+	}, nil
+}
+
+// WithRelease helps you set release/commit of currently running
+// code while initializing sentry client using function InitializeSentryClient
+func WithRelease(release string) func(*client) {
+	return func(c *client) {
+		c.c.SetRelease(release)
+	}
+}
+
+// WithEnvironment helps you set environment the deployed code is
+// running in while initializing sentry client using function
+// InitializeSentryClient
+func WithEnvironment(env string) func(*client) {
+	return func(c *client) {
+		c.c.SetEnvironment(env)
+	}
+}
+
+// waits on functions to be sent on channel
+// which are then executed
+func (c *client) loop() {
+	for op := range c.sendErr {
+		op()
+	}
+}
+
+// CaptureError sends error 'err' to Sentry, meanwhile also sets user
+// information by extracting user information from the context provided
+func (c *client) CaptureError(ctx context.Context, err error) {
+	// if method called during test which has uninitialized client
+	if c == nil {
+		return
+	}
+	// Extract user information. Ignoring error here but then before using the
+	// object user make sure to check if it wasn't nil.
+	user, _ := extractUserInfo(ctx)
```

Comment:
since user is a pointer and the function below is referring to it and not having it passed as an argument, wouldn't the function just be referring to the last user set?

Reply:
nope, it is not referring to last user set because every time a new pointer is created. The context of last call to `CaptureError` has nothing to do with the current call.

Reply:
Tried similar thing in sample code which shows no data will be overwritten: https://play.golang.org/p/UV2qLDJ5fwW

Reply:
Your example is only executing single threaded so that's not really representative. Just seeing the pointer ref, channel, and different goroutine got my mind thinking of how the for loop works.

Can't reproduce it locally either.